tests/unit/s2n_build_test.c

```diff
@@ -0,0 +1,150 @@
+/*
+* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+#define _GNU_SOURCE
+#include <openssl/crypto.h>
+#include <openssl/opensslv.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "s2n_test.h"
+
+/* s2n_test.h redefines OPEN_SSL_VERSION_NUMBER for LIBRESSL */
+#if defined(LIBRESSL_VERSION_NUMBER)
+const long unsigned openssl_version_number = 0x020000000L;
+#else
+const long unsigned openssl_version_number = OPENSSL_VERSION_NUMBER;
+#endif
+
+int main(int argc, char **argv)
+{
+    (void) argc, (void) argv;
+    BEGIN_TEST();
+
+    /*
+    * In the case that libcrypto is dynamically linked there is a chance that the
+    * build version and the linked version don't match. OPENSSL_VERSION_TEXT is
+    * a macro defined in openssl/crypto.h and should be returned from OpenSSL_version.
+    * In the event that we are dynamically linked to the wrong libcrypto the function,
+    * call will return a different version text than the one we used at a source level.
+    *
+    * If libcrypto is static linked, this is sure to be true.
+    */
+    if (SSLeay() != openssl_version_number) {
+        printf("\nOPENSSL_VERSION_NUMBER == %ld\n", (long int) openssl_version_number);
+        printf("SSLeay() == %ld\n", (long int) SSLeay());
+    }
+    EXPECT_EQUAL(SSLeay(), openssl_version_number);
```

Comment:
I'm probably misunderstanding something, but what does this test ensure exactly? If we have a header for one libcrypto, but we're dynamically linked with another libcrypto, wouldn't this test still think it's correct, since `SSLeay()` just returns the number in the header?